### PR TITLE
ARROW-10136: [Rust]: Fix null handling in StringArray and BinaryArray filtering, add BinaryArray::from_opt_vec

### PR DESCRIPTION
**Pull request by alamb**
When I use the `filter` kernel with Null strings, any input column that was Null turns into an empty string after filtering.

```
"foo"
"bar"
NULL
```
And the filter 
```
true
true
true
```

Will result in 
```
"foo"
"bar"
""
```

Rather than
```
"foo"
"bar"
NULL
```

It appears to work fine for primitive arrays (I'll comment inline).  I also added `BinaryArray::from_opt_vec` following the model of `PrimativeArray` and `StringArray` mostly so I could write a test.

### CODE REVIEW

File: rust/arrow/src/compute/kernels/filter.rs
Comment (alamb) [inline rust/arrow/src/compute/kernels/filter.rs:368]: Note using an `Option` is likely to increase the temporary storage requirements a bit.

It would likely be possible to avoid this allocation entirely if we used the lower level  `ArrayBuilder::with_bit_buffer`. 

I chose to follow the style of the rest of this module, though I would love opinions on trying to perf check this / optimize it (maybe a follow on JIRA ticket is enough)?

**Comment by nevi-me** [comment]
Thanks @alamb, this existed as https://issues.apache.org/jira/browse/ARROW-5352; so I'll close that out.

The same behaviour would occur with a `BinaryArray`, I haven't looked at this PR, but it's worthwhile to ensure that `StringArray` and `BinaryArray` are treated consistently.

And yes, the issue didn't affect primitive arrays, I think it was because we were pushing an empty string where we should have pushed a null slot on the null bitmap

File: rust/arrow/src/compute/kernels/filter.rs
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/filter.rs:368]: Doesn't an `Option` of a reference leverages null pointer optimization (on which `None` is represented by the null pointer, e.g. https://github.com/rust-lang/rust/issues/9378)?

IMO we should follow up on this: for kernels we have been using a mutable buffer with null masks as much as possible.

File: rust/arrow/src/array/array.rs
Comment (jorgecarleitao) [inline rust/arrow/src/array/array.rs:1270]: is this code tested somewhere?

File: rust/arrow/src/compute/kernels/filter.rs
Comment (jorgecarleitao) [inline rust/arrow/src/compute/kernels/filter.rs:693]: I suggest that we test the 3 quantities: `d.is_null(0)`, `d.value(0)`, `d.is_null(1)`. Alternatively, 

```rust
let expected = StringArray::from(vec![Some("hello"), None]);
assert_eq!(d, expected);
```

File: rust/arrow/src/compute/kernels/filter.rs
Comment (jorgecarleitao) [inline rust/arrow/src/compute/kernels/filter.rs:705]: same here.

**Review by jorgecarleitao — COMMENTED** [review]
Thanks a lot, @alamb . Looks good so far. I left some minor comments.

**Comment by alamb** [comment]
> The same behaviour would occur with a BinaryArray, I haven't looked at this PR, but it's worthwhile to ensure that StringArray and BinaryArray are treated consistently.

Thanks @nevi-me  -- yes I updated the code tried to handle both types in the same way (and including tests)

**Comment by alamb** [comment]
> Doesn't an Option of a reference leverages null pointer optimization (on which None is represented by the null pointer, e.g. rust-lang/rust#9378)?

@jorgecarleitao I believe you are correct:

This program:
```
fn main() {
    println!("The size of a &str is {}", std::mem::size_of::<&str>());
    println!("The size of an Option<&str> is {}", std::mem::size_of::<Option<&str>>());
}
```

Produces the following on my machine:
```
The size of a &str is 16
The size of an Option<&str> is 16
```

File: rust/arrow/src/array/array.rs
Comment (alamb) [reply rust/arrow/src/array/array.rs:1270]: It is tested (indirectly) in https://github.com/apache/arrow/pull/8303/files#diff-d7b0b7cde1850e8744ceda458c6dea81R700 -- but I think a more specific test would be valuable. I will add one.

File: rust/arrow/src/compute/kernels/filter.rs
Comment (alamb) [reply rust/arrow/src/compute/kernels/filter.rs:368]: > Doesn't an Option of a reference leverages null pointer optimization (on which None is represented by the null pointer, e.g. rust-lang/rust#9378)?

Yes, I believe you are correct.

This program:
```
fn main() {
    println!("The size of a &str is {}", std::mem::size_of::<&str>());
    println!("The size of an Option<&str> is {}", std::mem::size_of::<Option<&str>>());
}
```

Produces the following on my machine:
```
The size of a &str is 16
The size of an Option<&str> is 16
```

**Review by alamb — COMMENTED** [review]
@jorgecarleitao you said:

>IMO we should follow up on this: for kernels we have been using a mutable buffer with null masks as much as possible.

Can you perhaps let me know what you mean by this? Perhaps there is an example in the code you are thinking of?

File: rust/arrow/src/compute/kernels/filter.rs
Comment (alamb) [reply rust/arrow/src/compute/kernels/filter.rs:693]: I actually think checking `d.value(0)` (aka the value for the index of the null element) is probably not a good idea -- if an element is null the value shouldn't be checked. I think the `is_null(0)` check is a good one and will add that

File: rust/arrow/src/array/array.rs
Comment (alamb) [reply rust/arrow/src/array/array.rs:1270]: This turned out to be a great call @jorgecarleitao  -- I found a bug in this implementation while writing a test. Thank you for the suggestion. 💯

File: rust/arrow/src/compute/kernels/filter.rs
Comment (alamb) [reply rust/arrow/src/compute/kernels/filter.rs:356]: (this is the code for BinaryArray that @nevi-me  referred to in https://github.com/apache/arrow/pull/8303#issuecomment-701042100)

**Comment by alamb** [comment]
I think I have addressed all your (very helpful) comments @jorgecarleitao .

**Comment by jorgecarleitao** [comment]
> @jorgecarleitao you said:
> 
> > IMO we should follow up on this: for kernels we have been using a mutable buffer with null masks as much as possible.
> 
> Can you perhaps let me know what you mean by this? Perhaps there is an example in the code you are thinking of?

I am really, sorry, @alamb , I should have offered more context in the first place. :/

This in no way blocks this PR: IMO it is ready to merge if the relevant tests pass.

What I meant is that this code currently:

* creates `Vec<Option<T>>` through an iteration
* copies `Vec<Option<T>>` to the two buffers (when `from_opt_vec` is called)

it may be more efficient to create the buffers during the iteration, so that we avoid the copy (Vec -> buffers). In other words, the code in `from_opt_vec` could have been "injected" into the filter execution, where the `MuttableBuffer` and offsets and values buffer are created before the loop, and new elements are directly written to it. Does this any sense?

(as a side note, this is why I am proposing #8211 : IMO there is some boiler-plate copy-pasting to

1. initialize buffers
2. iterate
3. create `ArrayData` from buffers

which will continue to grow as we add more kernels, and whose pattern seems to be a `FromIter of fixed size`)

**Review by jorgecarleitao — APPROVED** [review]
LGTM, really nice additions.

**Comment by alamb** [comment]
@jorgecarleitao  -- yes thank you that makes a lot of sense. I have filed https://issues.apache.org/jira/browse/ARROW-10141 to track that

**Comment by alamb** [comment]
@andygrove  / @jorgecarleitao  / @nevi-me  I wonder if this PR might be merged anytime soon (I have a downstream project relying on this change)

The integration test failure https://github.com/apache/arrow/pull/8303/checks?check_run_id=1187275161 seems due to a network failure (not anything with this PR):
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.5.1:attach-descriptor (attach-descriptor) on project arrow-java-root: Execution attach-descriptor of goal org.apache.maven.plugins:maven-site-plugin:3.5.1:attach-descriptor failed: Plugin org.apache.maven.plugins:maven-site-plugin:3.5.1 or one of its dependencies could not be resolved: Could not transfer artifact org.apache.maven:maven-archiver:jar:2.5 from/to central (https://repo.maven.apache.org/maven2): Connection reset -> [Help 1]
```

**Comment by alamb** [comment]
I'll rebase to try and get a clean test run

**Comment by alamb** [comment]
All tests are passing now. 🎉

**Pull request closed without merging**
